After installing the German version of Cadenza, please import Testmappe_Ladesaeulen.zip arbeitsmappe, because some UI tests are based on this arbeitsmappe.
1. Klicke im Navigator auf die drei Punkte rechts neben der Suchleiste
2. Auf "Arbeitsmappe importieren" klicken
3. Bei "Datei *" auf "Bitte wählen Sie eine Datei aus." klicken und die angehängte ZIP-Datei auswählen
4. Wähle als Repository "Ladesaeulen"
5. Unten links auf "Importieren" klicken


In [1]:
import sqlite3
# Connect to the database
conn = sqlite3.connect('../data/raw/playwright_script.db')
cursor = conn.cursor()

res = cursor.execute("SELECT * FROM tests")
items = res.fetchall()

print( "there are {} data.".format(len(items)))

there are 100 data.


## UI Test Naming and Data Structure Guidelines

### 1. Naming Convention

To ensure traceability and organization within our UI tests, we have established a specific naming convention. Each test is named using a two-part numerical format, represented as `{x}.{y}`, where:

- `x` is the unique identifier for a set of UI tests.
- `y` indicates the step number within the UI test sequence, facilitating the tracking of test progress and specific actions.

This naming method allows for quick identification of specific test preconditions and steps.

### 2. Data Structure

Each UI test entry comprises six essential elements to ensure completeness and executability:

1. **ID**: Formatted as `x`.`y`, uniquely identifies each test and the step.
2. **Step Description**: Details the actions performed at each step, always starting from the first step.
3. **Expected Outcome**: Describes the expected result of the final step of the test.
4. **HTML File**: The HTML file generated after all actions have been executed, used for further analysis and review.
5. **Screenshot**: An image of the application or software interface taken after all actions have been completed, used for visual verification.
6. **Script**: Contains the Playwright code from the first to the last step, used to automate the UI test.


### 3. Data Examples

To better illustrate this data structure, here are two example entries:

In [2]:
items[0]

('1.1',
 '[1.1] Öffne die Arbeitsmappe "Übersicht Messstellen" im Ordner "Gewässergüte".',
 '[1.1] Expected result: Die Arbeitsmappe wird geöffnet, der Analysekontext ist nicht sichtbar.',
 '.\\html\\1_1.html',
 '.\\screenshot\\1_1.png',
 '.\\test_script\\1_1.spec.ts')

In [3]:
items[1]

('1.2',
 '[1.1] Öffne die Arbeitsmappe "Übersicht Messstellen" im Ordner "Gewässergüte". [1.2]  Öffnen der Tabellen-Sicht "Messstellenliste" über die Werkzeugliste der Arbeitsmappe.',
 '[1.2] Expected result: Die Tabelle "Messstelleninformationen" wird angezeigt, der Analysekontext ist sichtbar.',
 '.\\html\\1_2.html',
 '.\\screenshot\\1_2.png',
 '.\\test_script\\1_2.spec.ts')

In [2]:
items[0][4]

'.\\screenshot\\1_1.png'

To get previous or next id in order to search for the correct files as precondition or input:

In [20]:
current_id = items[0][4].split("\\")[-1].split(".")[0]
current_id

'1_1'

In [23]:
current_test, current_step = current_id.split("_")
next_id = f"{current_test}_{int(current_step)+1}"
previous_id = f"{current_test}_{int(current_step)-1}"

('1_0', '1_1', '1_2')

# Tipps : To complete the generation of playwright code for 1.2, 1.1 can be used as an initial condition (pre-condition)

## Task Description

### Objective
To utilize Large Language Models (LLM) for generating Playwright test scripts based on user descriptions. To achieve this goal, it is crucial to provide the LLM with sufficient context, including the software's structure, relevant HTML files, and interface screenshots.

### Methods and Suggestions

#### 1. How to Optimize Prompts?
- **Specificity and Clarity**: Ensure that prompts are as specific and clear as possible, clearly stating the specific actions to be automated and the expected outcomes.
- **Use Templates**: Develop a series of template prompts for common testing scenarios such as login, data entry, and functionality verification. These templates can help standardize inputs and reduce ambiguity.

#### 2. How to Encode HTML Files for LLM Understanding?
- **HTML Summarization**: Extract key information from HTML files (such as important IDs, class names, and tag types) instead of the entire HTML code, which helps the model understand the page structure.
- **DOM Structure Description**: Convert HTML files into descriptions of DOM tree structures, simplified into descriptions more akin to natural language, e.g., "The page contains a button with the id 'submit' inside a form."
- **Utilization of Tools and Libraries**: Use existing HTML parsing libraries such as BeautifulSoup to assist in processing and extracting data from HTML.

#### 3. How to Encode Image Information for LLM?
- **Image Description Generation**: Use computer vision models (like pretrained image recognition models) to generate descriptions of images, which are then used as textual inputs for the LLM.
- **Key Element Identification**: Identify and describe key UI elements in the images, such as buttons and input fields, and describe their relational positions, e.g., "A large search box is centered at the top of the page."
- **Use OCR**: For screenshots that contain text, employ Optical Character Recognition (OCR) technology to extract textual information, which may be crucial for understanding the interface.


# How to Run Playwright Code

Follow these detailed steps to set up your environment for running Playwright code using Visual Studio Code (VSCode).

## Step 1: Install Visual Studio Code

First, you need to install Visual Studio Code

- Visit the [Visual Studio Code download page](https://code.visualstudio.com/docs/?dv=win64user).
- Choose the appropriate installer for your operating system. This guide assumes you are using Windows 64-bit.
- Download and run the installer.
- Follow the on-screen instructions to complete the installation.

## Step 2: Install Node.js 

Playwright requires Node.js to run, so the next step is to install Node.js if it's not already installed on your system.

- Go to the [Node.js download page](https://nodejs.org/en/download/prebuilt-installer).    
- Download the installer appropriate for your operating system. For Windows, it's typically the Windows Installer (.msi).  Version !! V20.14.0
- Execute the downloaded file and follow the installation prompts. Make sure to include Node.js in your PATH if prompted.

## Step 3: Install Playwright in VSCode

After setting up VSCode and Node.js, you can install Playwright directly through VSCode.

- Open Visual Studio Code.
- Click on the Extensions view icon on the Sidebar or press `Ctrl+Shift+X`.
- In the Extensions search bar, type `Playwright` and look for the "Playwright Test for VSCode" extension.
- Click on the `Install` button to install the Playwright extension.
- After installation, press `Ctrl+Shift+P` to open the Command Palette.
- Type `Install Playwright` in the Command Palette and select the command that appears to install Playwright. This will set up Playwright with the necessary browser binaries.

## Step 4: Open and Run Your Playwright Script

Finally, you are ready to run your Playwright scripts.

- If necessary, restart Visual Studio Code to refresh the environment and ensure all extensions are properly loaded.
- Drag the playwright file script into tests fold.

By following these steps, you should be able to set up your development environment and run Playwright scripts smoothly within Visual Studio Code.
